From 91744bf811e9337a85e08843bbdb77ebf9cc89e0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Sebastian=20Dr=C3=B6ge?= <sebastian.droege@collabora.co.uk>
Date: Fri, 9 Mar 2012 08:44:13 +0100
Subject: [PATCH 02/13] Optionally revert to the old (pre-2.28) URI handler
 code

This is required to work with the GVfs distributed on Linux distributions
that ship with GLib 2.24 (e.g. Ubuntu Lucid and Debian Squeeze).
---
 gio/gdesktopappinfo.c | 51 +++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 51 insertions(+)

diff --git a/gio/gdesktopappinfo.c b/gio/gdesktopappinfo.c
index 35896b8..6fcd12f 100644
--- a/gio/gdesktopappinfo.c
+++ b/gio/gdesktopappinfo.c
@@ -4116,6 +4116,7 @@ out:
 GAppInfo *
 g_app_info_get_default_for_uri_scheme (const char *uri_scheme)
 {
+#ifndef GST_SDK_GLIB_GIO_OLD_URI_SCHEME_HANDLERS
   GAppInfo *app_info;
   char *content_type, *scheme_down;
 
@@ -4126,6 +4127,56 @@ g_app_info_get_default_for_uri_scheme (const char *uri_scheme)
   g_free (content_type);
 
   return app_info;
+#else
+  static gsize lookup = 0;
+  
+  if (g_once_init_enter (&lookup))
+    {
+      gsize setup_value = 1;
+      GDesktopAppInfoLookup *lookup_instance;
+      const char *use_this;
+      GIOExtensionPoint *ep;
+      GIOExtension *extension;
+      GList *l;
+ 
+      use_this = g_getenv ("GIO_USE_URI_ASSOCIATION");
+      
+      /* Ensure vfs in modules loaded */
+      _g_io_modules_ensure_loaded ();
+      
+      ep = g_io_extension_point_lookup (G_DESKTOP_APP_INFO_LOOKUP_EXTENSION_POINT_NAME);
+ 
+      lookup_instance = NULL;
+      if (use_this)
+	{
+	  extension = g_io_extension_point_get_extension_by_name (ep, use_this);
+	  if (extension)
+	    lookup_instance = g_object_new (g_io_extension_get_type (extension), NULL);
+	}
+      
+      if (lookup_instance == NULL)
+	{
+	  for (l = g_io_extension_point_get_extensions (ep); l != NULL; l = l->next)
+	    {
+	      extension = l->data;
+	      lookup_instance = g_object_new (g_io_extension_get_type (extension), NULL);
+	      if (lookup_instance != NULL)
+		break;
+	    }
+	}
+
+      if (lookup_instance != NULL)
+	setup_value = (gsize)lookup_instance;
+      
+      g_once_init_leave (&lookup, setup_value);
+    }
+
+  if (lookup == 1)
+    return NULL;
+
+  return g_desktop_app_info_lookup_get_default_for_uri_scheme (G_DESKTOP_APP_INFO_LOOKUP (lookup),
+							       uri_scheme);
+#endif
 }
 
 /* "Get all" API {{{2 */
-- 
2.7.4

